File: server/src/models/Equipment.ts
import mongoose, { Schema, Document } from 'mongoose';

export interface IEquipment extends Document {
  name: string;
  brand: string;
  equipmentModel: string;
  year: number;
  serialNumber: string;
  status: 'active' | 'maintenance' | 'blocked';
  qrCode?: string;
  documents: string[];
  checklistTemplateId?: mongoose.Types.ObjectId;
  lastMaintenance?: Date;
  nextMaintenance?: Date;
  createdAt: Date;
  updatedAt: Date;
}

const EquipmentSchema: Schema = new Schema(
  {
    name: { type: String, required: true },
    brand: { type: String, required: true },
    equipmentModel: { type: String, required: true },
    year: { type: Number, required: true },
    serialNumber: { type: String, required: true, unique: true },
    status: {
      type: String,
      enum: ['active', 'maintenance', 'blocked'],
      default: 'active',
    },
    qrCode: { type: String },
    documents: [{ type: String }],
    checklistTemplateId: { type: Schema.Types.ObjectId, ref: 'ChecklistTemplate' },
    lastMaintenance: { type: Date },
    nextMaintenance: { type: Date },
    imageUrl: { type: String },
  },
  { timestamps: true }
);

export default mongoose.model<IEquipment>('Equipment', EquipmentSchema);

File: mobile/lib/models/equipment.dart
import 'customer.dart';
class Equipment {
  final String id;
  final String name;
  final String brand;
  final String equipmentModel;
  final int year;
  final String serialNumber;
  final String status; // 'active', 'maintenance', 'blocked'
  final String? qrCode;
  final List<String> documents;
  final String? checklistTemplateId;
  final DateTime? lastMaintenance;
  final DateTime? nextMaintenance;
  final String? imageUrl;
  final double? capacity; // Added as per requirements

  Equipment({
    required this.id,
    required this.name,
    required this.brand,
    required this.equipmentModel,
    required this.year,
    required this.serialNumber,
    required this.status,
    this.qrCode,
    required this.documents,
    this.checklistTemplateId,
    this.lastMaintenance,
    this.nextMaintenance,
    this.imageUrl,
    this.capacity,
  });

  factory Equipment.fromJson(Map<String, dynamic> json) {
    return Equipment(
      id: json['_id'] ?? '',
      name: json['name'] ?? '',
      brand: json['brand'] ?? '',
      equipmentModel: json['equipmentModel'] ?? '',
      year: json['year'] ?? 0,
      serialNumber: json['serialNumber'] ?? '',
      status: json['status'] ?? 'active',
      qrCode: json['qrCode'],
      documents: List<String>.from(json['documents'] ?? []),
      checklistTemplateId: json['checklistTemplateId'],
      lastMaintenance: json['lastMaintenance'] != null ? DateTime.parse(json['lastMaintenance']) : null,
      nextMaintenance: json['nextMaintenance'] != null ? DateTime.parse(json['nextMaintenance']) : null,
      imageUrl: json['imageUrl'],
      capacity: json['capacity']?.toDouble(),
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is Equipment && runtimeType == other.runtimeType && id == other.id;

  @override
  int get hashCode => id.hashCode;

  /// Validates if equipment satisfies customer's required documentation and status
  bool isValidForCustomer(Customer customer) {
    // 1. Check status
    if (status != 'active') return false;

    // 2. Check maintenance
    if (nextMaintenance != null && nextMaintenance!.isBefore(DateTime.now())) {
      return false;
    }

    // 3. Check specific documentation required by customer
    for (final docId in customer.specificDocumentation) {
      if (!documents.contains(docId)) return false;
    }

    return true;
  }
}
